src/server/cdb/data/process.c

```diff
@@ -0,0 +1,368 @@
+        rsp->vstr.len = vstr->len;
+        rsp->vstr.data = vstr->data;
+
+        log_verb("found key at %p, location %p", key, vstr);
+        return true;
+    } else {
+        log_verb("key at %p not found", key);
+        return false;
+    }
+}
+
+static void
+_process_get(struct response *rsp, struct request *req)
+{
+    struct bstring *key;
+    struct response *r = rsp;
+    uint32_t i;
+
+    INCR(process_metrics, get);
+    /* use chained responses, move to the next response if key is found. */
+    for (i = 0; i < array_nelem(req->keys); ++i) {
+        INCR(process_metrics, get_key);
+        key = array_get(req->keys, i);
+        if (_get_key(r, key)) {
+            req->nfound++;
+            r->cas = false;
+            r = STAILQ_NEXT(r, next);
+            if (r == NULL) {
+                INCR(process_metrics, get_ex);
+                log_warn("get response incomplete due to lack of rsp objects");
+                return;
+            }
+            INCR(process_metrics, get_key_hit);
+        } else {
+            INCR(process_metrics, get_key_miss);
+        }
+    }
+    r->type = RSP_END;
+
+    log_verb("get req %p processed, %d out of %d keys found", req, req->nfound, i);
+}
+
+static void
+_process_gets(struct response *rsp, struct request *req)
+{
+    struct bstring *key;
+    struct response *r = rsp;
+    uint32_t i;
+
+    INCR(process_metrics, gets);
+    /* use chained responses, move to the next response if key is found. */
+    for (i = 0; i < array_nelem(req->keys); ++i) {
+        INCR(process_metrics, gets_key);
+        key = array_get(req->keys, i);
+        if (_get_key(r, key)) {
+            r->cas = true;
+            r = STAILQ_NEXT(r, next);
+            if (r == NULL) {
+                INCR(process_metrics, gets_ex);
+                log_warn("gets response incomplete due to lack of rsp objects");
+            }
+            req->nfound++;
+            INCR(process_metrics, gets_key_hit);
+        } else {
+            INCR(process_metrics, gets_key_miss);
+        }
+    }
+    r->type = RSP_END;
+
+    log_verb("gets req %p processed, %d out of %d keys found", req, req->nfound, i);
+}
+
+static void
+_process_invalid(struct response *rsp, struct request *req)
+{
+    INCR(process_metrics, invalid);
+    rsp->type = RSP_CLIENT_ERROR;
+    rsp->vstr = str2bstr(CMD_ERR_MSG);
+    log_verb("req %p processed, rsp type %d", req, rsp->type);
+}
+
+static void
+_process_flush(struct response *rsp, struct request *req)
+{
+    INCR(process_metrics, flush);
+    rsp->type = RSP_CLIENT_ERROR;
+    rsp->vstr = str2bstr(CMD_ERR_MSG);
+    log_info("flush req %p processed, rsp type %d", req, rsp->type);
+}
+
+void
+process_request(struct response *rsp, struct request *req)
+{
+    log_verb("processing req %p, write rsp to %p", req, rsp);
+    INCR(process_metrics, process_req);
+
+    switch (req->type) {
+    case REQ_GET:
+        _process_get(rsp, req);
+        break;
+
+    case REQ_GETS:
```

Comment:
I think this can also be moved to the invalid section, `gets` is used for  subsequent`cas`